### PR TITLE
Handle `file://` in specs of junit reporter - fixes #9352

### LINKED ISSUE #9352: [🐛 Bug]: wdio 8 junit Report is empty when running Cucumber tests
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

8.0.9

### Node.js Version

18.12.1

### Mode

WDIO Testrunner

### Which capabilities are you using?

_No response_

### What happened?

Junit reported is empty when running with Cucumber.

### What is your expected behavior?

junit report is populated with tests results on version 8, as on version 7.25.1.

### How to reproduce the bug.

Trying out new wdio version 8.

Started the project as described on webdriver.io website.

**Command executed:**
`npm init wdio ./newWdio8`

**package.json**
```
{
  "name": "webdriverio-tests",
  "version": "0.1.0",
  "private": true,
  "devDependencies": {
    "@wdio/cli": "^8.0.9",
    "@wdio/cucumber-framework": "^8.0.9",
    "@wdio/junit-reporter": "^8.0.8",
    "@wdio/local-runner": "^8.0.9",
    "@wdio/spec-reporter": "^8.0.8",
    "chromedriver": "^108.0.0",
    "wdio-chromedriver-service": "^8.0.1"
  },
  "scripts": {
    "wdio": "wdio run wdio.conf.js"
  }
}
```

**wdio.conf.js**
```
exports.config = {
  //
  // ====================
  // Runner Configuration
  // ====================
  // WebdriverIO supports running e2e tests as well as unit and component tests.
  runner: "local",

  //
  // ==================
  // Specify Test Files
  // ==================
  // Define which test specs should run. The pattern is relative to the directory
  // from which `wdio` was called.
  //
  // The specs are defined as an array of spec files (optionally using wildcards
  // that will be expanded). The test for each spec file will be run in a separate
  // worker process. In order to have a group of spec files run in the same worker
  // process simply enclose them in an array within the specs array.
  //
  // If you are calling `wdio` from an NPM script (see https://docs.npmjs.com/cli/run-script),
  // then the current working directory is where your `package.json` resides, so `wdio`
  // will be called from there.
  //
  specs: ["./features/**/*.feature"],
  // Patterns to exclude.
  exclude: [
    // 'path/to/excluded/files'
  ],
  //
  // ============
  // Capabilities
  // ============
  // Define your capabilities here. WebdriverIO can run multiple capabilities at the same
  // time. Depending on the number of capabilities, WebdriverIO launches several test
  // sessions. Within your capabilities you can overwrite the spec and exclude options in
  // order to group specific specs to a specific capability.
  //
  // First, you can define how many instances should be started at the same time. Let's
  // say you have 3 different capabilities (Chrome, Firefox, and Safari) and you have
  // set maxInstances to 1; wdio will spawn 3 processes. Therefore, if you have 10 spec
  // files and you set maxInstances to 10, all spec files will get tested at the same time
  // and 30 processes will get spawned. The property handles how many capabilities
  // from the same test should run tests.
  //
  maxInstances: 10,
  //
  // If you have trouble getting all important capabilities together, check out the
  // Sauce Labs platform configurator - a great tool to configure your capabilities:
  // https://saucelabs.com/platform/platform-configurator
  //
  capabilities: [
    {
      // maxInstances can get overwritten per capability. So if you have an in-house Selenium
      // grid with only 5 firefox instances available you can make sure that not more than
      // 5 instances get started at a time.
      maxInstances: 5,
      //
      browserName: "chrome",
      acceptInsecureCerts: true,
      // If outputDir is provided WebdriverIO can capture driver session logs
      // it is possible to configure which logTypes to include/exclude.
      // excludeDriverLogs: ['*'], // pass '*' to exclude all driver session logs
      // excludeDriverLogs: ['bugreport', 'server'],
    },
  ],
  //
  // ===================
  // Test Configurations
  // ===================
  // Define all options that are relevant for the WebdriverIO instance here
  //
  // Level of logging verbosity: trace | debug | info | warn | error | silent
  logLevel: "info",
  //
  // Set specific log levels per logger
  // loggers:
  // - webdriver, webdriverio
  // - @wdio/browserstack-service, @wdio/devtools-service, @wdio/sauce-service
  // - @wdio/mocha-framework, @wdio/jasmine-framework
  // - @wdio/local-runner
  // - @wdio/sumologic-reporter
  // - @wdio/cli, @wdio/config, @wdio/utils
  // Level of logging verbosity: trace | debug | info | warn | error | silent
  // logLevels: {
  //     webdriver: 'info',
  //     '@wdio/appium-service': 'info'
  // },
  //
  // If you only want to run your tests until a specific amount of tests have failed use
  // bail (default is 0 - don't bail, run all tests).
  bail: 0,
  //
  // Set a base URL in order to shorten url command calls. If your `url` parameter starts
  // with `/`, the base url gets prepended, not including the path portion of your baseUrl.
  // If your `url` parameter starts without a scheme or `/` (like `some/path`), the base url
  // gets prepended directly.
  baseUrl: "http://localhost",
  //
  // Default timeout for all waitFor* commands.
  waitforTimeout: 10000,
  //
  // Default timeout in milliseconds for request
  // if browser driver or grid doesn't send response
  connectionRetryTimeout: 120000,
  //
  // Default request retries count
  connectionRetryCount: 3,
  //
  // Test runner services
  // Services take over a specific job you don't want to take care of. They enhance
  // your test setup with almost no effort. Unlike plugins, they don't add new
  // commands. Instead, they hook themselves up into the test process.
  services: ["chromedriver"],

  // Framework you want to run your specs with.
  // The following are supported: Mocha, Jasmine, and Cucumber
  // see also: https://webdriver.io/docs/frameworks
  //
  // Make sure you have the wdio adapter package for the specific framework installed
  // before running any tests.
  framework: "cucumber",
  //
  // The number of times to retry the entire specfile when it fails as a whole
  // specFileRetries: 1,
  //
  // Delay in seconds between the spec file retry attempts
  // specFileRetriesDelay: 0,
  //
  // Whether or not retried specfiles should be retried immediately or deferred to the end of the queue
  // specFileRetriesDeferred: false,
  //
  // Test reporter for stdout.
  // The only one supported by default is 'dot'
  // see also: https://webdriver.io/docs/dot-reporter
  reporters: ["spec", "junit"],

  //
  // If you are using Cucumber you need to specify the location of your step definitions.
  cucumberOpts: {
    // <string[]> (file/dir) require files before executing features
    require: ["./features/step-definitions/steps.js"],
    // <boolean> show full backtrace for errors
    backtrace: false,
    // <string[]> ("extension:module") require files with the given EXTENSION after requiring MODULE (repeatable)
    requireModule: [],
    // <boolean> invoke formatters without executing steps
    dryRun: false,
    // <boolean> abort the run on first failure
    failFast: false,
    // <boolean> hide step definition snippets for pending steps
    snippets: true,
    // <boolean> hide source uris
    source: true,
    // <boolean> fail if there are any undefined or pending steps
    strict: false,
    // <string> (expression) only execute the features or scenarios with tags matching the expression
    tagExpression: "",
    // <number> timeout for step definitions
    timeout: 60000,
    // <boolean> Enable this config to treat undefined definitions as warnings.
    ignoreUndefinedDefinitions: false,
  },

  //
  // =====
  // Hooks
  // =====
  // WebdriverIO provides several hooks you can use to interfere with the test process in order to enhance
  // it and to build services around it. You can either apply a single function or an array of
  // methods to it. If one of them returns with a promise, WebdriverIO will wait until that promise got
  // resolved to continue.
  /**
   * Gets executed once before all workers get launched.
   * @param {Object} config wdio configuration object
   * @param {Array.<Object>} capabilities list of capabilities details
   */
  // onPrepare: function (config, capabilities) {
  // },
  /**
   * Gets executed before a worker process is spawned and can be used to initialise specific service
   * for that worker as well as modify runtime environments in an async fashion.
   * @param  {String} cid      capability id (e.g 0-0)
   * @param  {[type]} caps     object containing capabilities for session that will be spawn in the worker
   * @param  {[type]} specs    specs to be run in the worker process
   * @param  {[type]} args     object that will be merged with the main configuration once worker is initialized
   * @param  {[type]} execArgv list of string arguments passed to the worker process
   */
  // onWorkerStart: function (cid, caps, specs, args, execArgv) {
  // },
  /**
   * Gets executed just after a worker process has exited.
   * @param  {String} cid      capability id (e.g 0-0)
   * @param  {Number} exitCode 0 - success, 1 - fail
   * @param  {[type]} specs    specs to be run in the worker process
   * @param  {Number} retries  number of retries used
   */
  // onWorkerEnd: function (cid, exitCode, specs, retries) {
  // },
  /**
   * Gets executed just before initialising the webdriver session and test framework. It allows you
   * to manipulate configurations depending on the capability or spec.
   * @param {Object} config wdio configuration object
   * @param {Array.<Object>} capabilities list of capabilities details
   * @param {Array.<String>} specs List of spec file paths that are to be run
   * @param {String} cid worker id (e.g. 0-0)
   */
  // beforeSession: function (config, capabilities, specs, cid) {
  // },
  /**
   * Gets executed before test execution begins. At this point you can access to all global
   * variables like `browser`. It is the perfect place to define custom commands.
   * @param {Array.<Object>} capabilities list of capabilities details
   * @param {Array.<String>} specs        List of spec file paths that are to be run
   * @param {Object}         browser      instance of created browser/device session
   */
  // before: function (capabilities, specs) {
  // },
  /**
   * Runs before a WebdriverIO command gets executed.
   * @param {String} commandName hook command name
   * @param {Array} args arguments that command would receive
   */
  // beforeCommand: function (commandName, args) {
  // },
  /**
   * Cucumber Hooks
   *
   * Runs before a Cucumber Feature.
   * @param {String}                   uri      path to feature file
   * @param {GherkinDocument.IFeature} feature  Cucumber feature object
   */
  // beforeFeature: function (uri, feature) {
  // },
  /**
   *
   * Runs before a Cucumber Scenario.
   * @param {ITestCaseHookParameter} world    world object containing information on pickle and test step
   * @param {Object}                 context  Cucumber World object
   */
  // beforeScenario: function (world, context) {
  // },
  /**
   *
   * Runs before a Cucumber Step.
   * @param {Pickle.IPickleStep} step     step data
   * @param {IPickle}            scenario scenario pickle
   * @param {Object}             context  Cucumber World object
   */
  // beforeStep: function (step, scenario, context) {
  // },
  /**
   *
   * Runs after a Cucumber Step.
   * @param {Pickle.IPickleStep} step             step data
   * @param {IPickle}            scenario         scenario pickle
   * @param {Object}             result           results object containing scenario results
   * @param {boolean}            result.passed    true if scenario has passed
   * @param {string}             result.error     error stack if scenario failed
   * @param {number}             result.duration  duration of scenario in milliseconds
   * @param {Object}             context          Cucumber World object
   */
  // afterStep: function (step, scenario, result, context) {
  // },
  /**
   *
   * Runs after a Cucumber Scenario.
   * @param {ITestCaseHookParameter} world            world object containing information on pickle and test step
   * @param {Object}                 result           results object containing scenario results
   * @param {boolean}                result.passed    true if scenario has passed
   * @param {string}                 result.error     error stack if scenario failed
   * @param {number}                 result.duration  duration of scenario in milliseconds
   * @param {Object}                 context          Cucumber World object
   */
  // afterScenario: function (world, result, context) {
  // },
  /**
   *
   * Runs after a Cucumber Feature.
   * @param {String}                   uri      path to feature file
   * @param {GherkinDocument.IFeature} feature  Cucumber feature object
   */
  // afterFeature: function (uri, feature) {
  // },

  /**
   * Runs after a WebdriverIO command gets executed
   * @param {String} commandName hook command name
   * @param {Array} args arguments that command would receive
   * @param {Number} result 0 - command success, 1 - command error
   * @param {Object} error error object if any
   */
  // afterCommand: function (commandName, args, result, error) {
  // },
  /**
   * Gets executed after all tests are done. You still have access to all global variables from
   * the test.
   * @param {Number} result 0 - test pass, 1 - test fail
   * @param {Array.<Object>} capabilities list of capabilities details
   * @param {Array.<String>} specs List of spec file paths that ran
   */
  // after: function (result, capabilities, specs) {
  // },
  /**
   * Gets executed right after terminating the webdriver session.
   * @param {Object} config wdio configuration object
   * @param {Array.<Object>} capabilities list of capabilities details
   * @param {Array.<String>} specs List of spec file paths that ran
   */
  // afterSession: function (config, capabilities, specs) {
  // },
  /**
   * Gets executed after all workers got shut down and the process is about to exit. An error
   * thrown in the onComplete hook will result in the test run failing.
   * @param {Object} exitCode 0 - success, 1 - fail
   * @param {Object} config wdio configuration object
   * @param {Array.<Object>} capabilities list of capabilities details
   * @param {<Object>} results object containing test results
   */
  // onComplete: function(exitCode, config, capabilities, results) {
  // },
  /**
   * Gets executed when a refresh happens.
   * @param {String} oldSessionId session ID of the old session
   * @param {String} newSessionId session ID of the new session
   */
  // onReload: function(oldSessionId, newSessionId) {
  // }
};
```

**Executed command:**
` npm run wdio`

**Output results in terminal:**
```
[chrome 108.0.5359.94 mac os x #0-0] Running: chrome (v108.0.5359.94) on mac os x
[chrome 108.0.5359.94 mac os x #0-0] Session ID: abda0f2b26484605107181615296f867
[chrome 108.0.5359.94 mac os x #0-0]
[chrome 108.0.5359.94 mac os x #0-0] » /features/login.feature
[chrome 108.0.5359.94 mac os x #0-0] The Internet Guinea Pig Website
[chrome 108.0.5359.94 mac os x #0-0] As a user, I can log into the secure area
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Given I am on the login page
[chrome 108.0.5359.94 mac os x #0-0]    ✓ When I login with tomsmith and SuperSecretPassword!
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Then I should see a flash message saying You logged into a secure area!
[chrome 108.0.5359.94 mac os x #0-0]
[chrome 108.0.5359.94 mac os x #0-0] As a user, I can log into the secure area
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Given I am on the login page
[chrome 108.0.5359.94 mac os x #0-0]    ✓ When I login with foobar and barfoo
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Then I should see a flash message saying Your username is invalid!
[chrome 108.0.5359.94 mac os x #0-0]
[chrome 108.0.5359.94 mac os x #0-0] 6 passing (3.4s)


 "junit" Reporter:
<?xml version="1.0" encoding="UTF-8"?>
<testsuites/>

Spec Files:      1 passed, 1 total (100% completed) in 00:00:07 
```
**junit Reporter is empty.**

**Note**

Trying the same on previous version 7.25.1 and got the outcome:

```
[chrome 108.0.5359.94 mac os x #0-0] Running: chrome (v108.0.5359.94) on mac os x
[chrome 108.0.5359.94 mac os x #0-0] Session ID: 55c776c0b3b2b7fd421cd1f9442e7e9f
[chrome 108.0.5359.94 mac os x #0-0]
[chrome 108.0.5359.94 mac os x #0-0] » /features/login.feature
[chrome 108.0.5359.94 mac os x #0-0] The Internet Guinea Pig Website
[chrome 108.0.5359.94 mac os x #0-0] As a user, I can log into the secure area
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Given I am on the login page
[chrome 108.0.5359.94 mac os x #0-0]    ✓ When I login with tomsmith and SuperSecretPassword!
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Then I should see a flash message saying You logged into a secure area!
[chrome 108.0.5359.94 mac os x #0-0]
[chrome 108.0.5359.94 mac os x #0-0] As a user, I can log into the secure area
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Given I am on the login page
[chrome 108.0.5359.94 mac os x #0-0]    ✓ When I login with foobar and barfoo
[chrome 108.0.5359.94 mac os x #0-0]    ✓ Then I should see a flash message saying Your username is invalid!
[chrome 108.0.5359.94 mac os x #0-0]
[chrome 108.0.5359.94 mac os x #0-0] 6 passing (3.6s)


 "junit" Reporter:
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="The Internet Guinea Pig Website" timestamp="2022-12-08T10:49:06" time="3.569" tests="2" failures="0" errors="0" skipped="0">
    <properties>
      <property name="specId" value="0"/>
      <property name="featureName" value="The Internet Guinea Pig Website"/>
      <property name="capabilities" value="chrome.108_0_5359_94.macosx"/>
      <property name="featureFile" value="./features/login.feature"/>
    </properties>
    <testcase classname="CucumberJUnitReport-chrome.108_0_5359_94.macosx.The Internet Guinea Pig Website" name="As a user I can log into the secure area" time="2.796">
      <system-out><![CDATA[
✅ Given I am on the login page✅ When I login with tomsmith and SuperSecretPassword!✅ Then I should see a flash message saying You logged into a secure area!
]]></system-out>
    </testcase>
    <testcase classname="CucumberJUnitReport-chrome.108_0_5359_94.macosx.The Internet Guinea Pig Website" name="As a user I can log into the secure area" time="0.759">
      <system-out><![CDATA[
✅ Given I am on the login page✅ When I login with foobar and barfoo✅ Then I should see a flash message saying Your username is invalid!
]]></system-out>
    </testcase>
  </testsuite>
</testsuites>

Spec Files:      1 passed, 1 total (100% completed) in 00:00:08 
```
```
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues
```

### PR DESCRIPTION
**Pull request by nextlevelbeard**
## Proposed changes

 Handle file:// in specs of junit reporter - fixes #9352 
 
[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)
- [X] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by nextlevelbeard** [comment]
> As suggested we should work with `url.fileURLToPath()` here

Attempting to use `url.fileURLToPath` on a relative path such as `'./features/sample_feature.feature'` on macOS leads to `TypeError: File URL host must be "localhost" or empty on darwin`

Best to simply remove the prefix for the comparison.

**Review by christian-bromann — APPROVED** [review]
I guess we don't need to make it more complicated as it is, LGTM 👍

File: packages/wdio-junit-reporter/tests/reporter.test.ts
Comment (christian-bromann) [inline packages/wdio-junit-reporter/tests/reporter.test.ts:143]: Minor formatting
```suggestion
        reporter = new WDIOJunitReporter({ stdout: true })
        reporter.suites = { ...featuresLog } as any
        const runner = { ...cucumberRunnerLog }
        runner.specs = runner.specs.map(s => `file://${s}`)

        // verifies the content of the report but omits format by stripping all whitespace and new lines
        expect(reporter['_buildJunitXml'](runner as any).replace(/\s/g, '')).toMatchSnapshot()
    })

    it('generates xml output (Cucumber-style) when feature has file:// protocol and the addFileAttribute option is set', () => {
        reporter = new WDIOJunitReporter({ stdout: true, addFileAttribute: true })
        reporter.suites = { ...featuresLog } as any
        const runner = { ...cucumberRunnerLog }
        runner.specs = runner.specs.map(s => `file://${s}`)

        // verifies the content of the report but omits format by stripping all whitespace and new lines
        expect(reporter['_buildJunitXml'](runner as any).replace(/\s/g, '')).toMatchSnapshot()
```

**Comment by nextlevelbeard** [comment]
No write access, ship it at your will

**Pull request merged**
